kubelet-to-gcm/monitor/kubelet_to_gcm.go

```diff
@@ -0,0 +1,412 @@
+func NewTranslator(zone, project, cluster, instanceID string, resolution time.Duration) *Translator {
+	return &Translator{
+		zone:       zone,
+		project:    project,
+		cluster:    cluster,
+		instanceID: instanceID,
+		resolution: resolution,
+	}
+}
+
+// Translate translates a summary to its TimeSeries.
+func (t *Translator) Translate(summary *stats.Summary) (*v3.CreateTimeSeriesRequest, error) {
+	var ts []*v3.TimeSeries
+	nodeTs, err := t.translateNode(summary.Node)
+	if err != nil {
+		return nil, err
+	}
+	podsTs, err := t.translateContainers(summary.Pods)
+	if err != nil {
+		return nil, err
+	}
+	ts = append(ts, nodeTs...)
+	ts = append(ts, podsTs...)
+	return &v3.CreateTimeSeriesRequest{TimeSeries: ts}, nil
+}
+
+func (t *Translator) translateNode(node stats.NodeStats) ([]*v3.TimeSeries, error) {
+	var timeSeries []*v3.TimeSeries
+
+	monitoredLabels := map[string]string{
+		"project_id":     t.project,
+		"cluster_name":   t.cluster,
+		"zone":           t.zone,
+		"instance_id":    t.instanceID,
+		"namespace_id":   "",
+		"pod_id":         "machine",
+		"container_name": "",
+	}
+	tsFactory := newTimeSeriesFactory(monitoredLabels, t.resolution)
+
+	// Uptime. This is embedded: there's no nil check.
+	now := time.Now()
+	uptimePoint := &v3.Point{
+		Interval: &v3.TimeInterval{
+			EndTime:   now.Format(time.RFC3339),
+			StartTime: now.Add(-1 * t.resolution).Format(time.RFC3339),
+		},
+		Value: &v3.TypedValue{
+			DoubleValue: float64(time.Since(node.StartTime.Time).Seconds()),
+		},
+	}
+	timeSeries = append(timeSeries, tsFactory.newTimeSeries(noLabels, uptimeMD, uptimePoint))
+
+	// Memory stats.
+	memTS, err := translateMemory(node.Memory, tsFactory)
+	if err != nil {
+		return nil, err
+	}
+	timeSeries = append(timeSeries, memTS...)
+
+	// File-system stats.
+	fsTS, err := translateFS("/", node.Fs, tsFactory)
+	if err != nil {
+		return nil, err
+	}
+	timeSeries = append(timeSeries, fsTS...)
+
+	// CPU stats.
+	cpuTS, err := translateCPU(node.CPU, tsFactory)
+	if err != nil {
+		return nil, err
+	}
+	timeSeries = append(timeSeries, cpuTS...)
+
+	return timeSeries, nil
+}
+
+func (t *Translator) translateContainers(pods []stats.PodStats) ([]*v3.TimeSeries, error) {
+	var timeSeries []*v3.TimeSeries
+	metricsSeen := make(map[string]time.Time)
+	metrics := make(map[string][]*v3.TimeSeries)
+
+	for _, pod := range pods {
+		namespace := pod.PodRef.Namespace
+		podID := pod.PodRef.Name
+		// There can be duplicate data points for containers, so only
+		// take the latest one.
+		for _, container := range pod.Containers {
+			containerName := container.Name
+			// Check for duplicates
+			if container.StartTime.Time.Before(metricsSeen[containerName]) || container.StartTime.Time.Equal(metricsSeen[containerName]) {
+				continue
+			}
+			metricsSeen[containerName] = container.StartTime.Time
+			var containerSeries []*v3.TimeSeries
+
+			monitoredLabels := map[string]string{
```

Comment:
I believe that Heapster pushes a bit more metrics (like image name). Are the names in sync with heapster?